packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,83 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL, SLOT_COMPONENT_TYPEOF_SYMBOL } from './constants';
+import * as React from 'react';
+import { SlotPropsRecord } from './types';
+import { ExtractSlotProps } from './types';
+import { PropsWithoutRef } from './types';
+import { getNativeElementProps } from '../utils/getNativeElementProps';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  elementType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+};
+
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  options: { shorthand: Props | SlotShorthandValue | undefined; required: true } & SlotOptions<Props>,
```

Comment:
The "shorthand" argument is not necessarily a shorthand value. Not sure what the best name would be. Perhaps `value`, which is the name that `resolveShorthand` uses.

Also, super-nit, the object should be called "params" or something (and the type "SlotParams"), rather than "options"/"SlotOptions" since they are not optional 🙂.

```suggestion
  params: { value: Props | SlotShorthandValue | undefined; required: true } & SlotParams<Props>,
```